docs: add `x/posts` specs (#911)
desmos-labs/desmos: x/posts/spec/02-concepts.md

#### TextTag
Both hashtags and mentions are represented as a `TextTag`. The `TextTag` structure contains the necessary fields that ease the process of 
text's parsing.

##### Start
Index within the post text at which the tag starts.

##### End
Index within the post text at which the tag ends.

##### Tag
The actual value of the tag. Usually this is going to be either the hashtag value, or the address of the mentioned user (if within the post text the DTag is used to reference the user).

#### Url
Within Desmos there is no URL validation, so this type can be used to represent any kind of URL that might be present inside the post text.

##### Start
Index within the post text at which the URL starts.

##### End
Index within the post text at which the URL ends.

##### Url
The actual value of the url where the user should be redirected to.

##### Display url (Optional)
Value to be displayed within the post text instead of the URL value itself. This might be useful if you want to shorten a link or if you want to represent a clickable text.

> 

Reviewer: ```suggestion

### Tags (optional)
Tags represent a set of strings that can be useful to quickly identify the post's contents. They can be used for content categorization or other different scopes.

```

---
id: concepts
title: Concepts
sidebar_label: Concepts
slug: concepts
---

# Concepts 

## Post
A Post is a structure representing any kind of content in a social network. It can contain some text and be enriched with:
- Attachments such as media (pictures, clips, videos, GIFs) and polls;
- Links;
- Mentions of users;
- Hashtags values;
- Quotes of other posts.

### Subspace ID
The [subspace] ID indicates the ID of the Dapp where the post is hosted and lives.

### Section ID 
The [section] ID indicates the ID of the subspace's section where the post lives.

### ID
The unique ID that identifies the post itself. This ID is automatically assigned to the post at the moment of its 
creation in a sequential way (e.g. if there's 4 posts in the chain, the one we are creating will have id equal to 5).

### External ID (Optional)

> External ID indicates and external ID attached to the post. It is represented by a string. //TODO add some more info 

Reviewer 2: Can the TODO reference be removed?

### Text (Optional)
The text is the actual textual content of the post. It has a fixed max length that is determined by an on-chain governance parameter.

### Entities (Optional)
Entities are particular parts of the text that can be parsed out of it in order to be displayed in custom ways.
Entities are divided in 3 different categories:
- Hashtags (i.e. #desmos)
- Mentions (i.e. @desmos1xcfui...., @Forbole)
- Urls (i.e. https://desmos.network, ftp://user:password@example.com/image.png)

> #### Tag

Reviewer 2: Should be TextTag.

The actual value of the url where the user should be redirected to.

##### Display url (Optional)
The visual url that should be displayed in the app. Ideally something that shorten or make an url more human-readable.

### Author
The address of the author of the post.

### Conversation ID (optional)
The ID of the original post of the conversation. This ID identifies the parent post (if existent) where the conversation
started.

### Referenced Posts
Referenced posts are represented by a `PostReference` structure.

#### PostReference
A reference to an external post, usable for reply, repost or quote purposes.

##### Type
This is the type of the reference. It can be one of the following values:

```js reference
https://github.com/desmos-labs/desmos/blob/14137abeaf79d61c92bc805d2f93f9f0e6e99138/proto/desmos/posts/v1/models.proto#L70
```

##### Post ID
The ID of the referenced post.

##### Position
This field should be used only when the `Type` field is equal to `TYPE_QUOTE` to indicates where the `PostReference`
start in the post's text.

### Reply Setting
This field contains the possible reply settings that a post can have. It can be one of the following values:

```js reference 

> https://github.com/desmos-labs/desmos/blob/14137abeaf79d61c92bc805d2f93f9f0e6e99138/proto/desmos/posts/v1/models.proto#L98

Reviewer 2: as above, would be better to be shown as a form.

```

### Creation Date
The creation date of the post.

### Last Edited Date
The las time the post has been edited.

## Attachment
An attachment represent something that can be added to a post in order to enrich it or give it some additional use cases.

### Subspace ID
The [subspace] ID indicates the ID of the Dapp where the attachment is hosted and lives.

> ### Section ID

Reviewer 2: It is removed by #953

### Provided answers
The possible answers choices the users have to reply the poll. They are represented by a `ProvidedAnswer` structure.

#### Provided Answer
The representation of a provided answer for a poll.

##### Text (Optional)
The text of the answer.

##### Attachments (Optional)
The [attachments](#attachment) of the answer. If not provided, a text answer has to be specified.

### End Date
The date when the poll will close.

### Allow Multiple Answers
This field tells if the poll allows multiple answers or not.

### Allow Answer Edits
This field tells if the poll allows users to edit their answers or not.

### Final Tally Results 
This fields contains the final results of the poll.

#### Results
The answers' results represent by the `AnswerResult` structure.

##### Answer Result
This field contains the result of a single poll provided answer

###### Answer Index
The Index of the answer inside the `ProvidedAnswer`s array.

###### Votes
The number of votes received by an answer.

## User Answer
The user answer represent an answer given by a user to a poll.

#### Subspace ID
The [subspace] ID indicates the ID of the Dapp where the user answer is hosted and lives.

> #### Section ID

Reviewer 2: as above

##### Post ID
The ID of the referenced post.

##### Position
This field should be used only when the `Type` field is equal to `TYPE_QUOTE` to indicates where the `PostReference`
start in the post's text.

### Reply Setting
This field contains the possible reply settings that a post can have. It can be one of the following values:

| **Name**                    | **Description**                                              |  
|:----------------------------|:-------------------------------------------------------------|
| `REPLY_SETTING_UNSPECIFIED` | No reference specified                                       |
| `REPLY_SETTING_EVERYONE`    | This reference represents a reply to the specified post      |
| `REPLY_SETTING_FOLLOWERS`   | This reference represents a quote of the specified post      |
| `REPLY_SETTING_MUTUAL`      | This reference represents a repost of the specified post     |
| `REPLY_SETTING_MENTIONS`    | Only people mentioned inside this post will be able to reply |

### Creation Date
The creation date of the post.

### Last Edited Date
The las time the post has been edited.

## Attachment

> An attachment represent something that can be added to a post in order to enrich it or give it some additional use cases.

Reviewer: ```suggestion
An attachment represents any kind of media that can be attached to a post, such as (but not limited to): an image, a GIF, or a poll.
```

### Subspace ID

> The [subspace] ID indicates the ID of the Dapp where the attachment is hosted and lives.

Reviewer: ```suggestion
The subspace id specifies the id of the subspace inside which the post related to the attachment is.
```

### Post ID

> The [post](#Post) ID to which the attachment is linked.

Reviewer: ```suggestion
The post id specifies the id of the post to which the attachment is associated.
```

### ID

> The unique ID that identifies the attachment. This ID is automatically assigned the same way the post one is.

Reviewer: ```suggestion
An attachment id is the unique identifier within a specific post. This is automatically created when handling a `MsgCreatePost` or `MsgAddPostAttachment`. With the subspace id and post id, it can be used to uniquely identify an attachment within Desmos.
```

### Content

> The content of the attachment. It can be either:
> - A media content (pictures, clips, videos, GIFs);
> - A poll.

Reviewer: ```suggestion
The content of the attachment. It can be either a media content (e.g. a picture, a video, a GIF, etc), or a poll.
```

## Media

> The Media structure represent a media content of any kind from pictures, to clips, videos, GIFs, etc... attached to 
> a post.

Reviewer: ```suggestion
A media content can be used to specify the details of a media attachment (e.g. an image, a video, a GIF, etc).
```

### URI

> The URI address referencing the media.

Reviewer: ```suggestion
The URI referencing the media.
```

### Mime type
The [mime type](https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/MIME_types) of the media.

## Poll

> The Poll structure represents a poll attached to a post.

Reviewer: ```suggestion
A poll media can be used to represent a poll attachment.
```

### Question

> The question field define the question of the poll.

Reviewer: ```suggestion
The question of the poll.
```

### Provided answers

> The possible answers choices the users have to reply the poll. They are represented by a `ProvidedAnswer` structure.

Reviewer: ```suggestion
The possible answers the users can chose from to reply the poll. They are each an instance of `ProvidedAnswer`.
```

#### Provided Answer
The representation of a provided answer for a poll.

##### Text (Optional)

> The text of the answer.

Reviewer: ```suggestion
The text of the answer. Note that either one of the text or attachments field must be specified within a provided answer to make sure it's valid.
```

##### Attachments (Optional)

> The [attachments](#attachment) of the answer. If not provided, a text answer has to be specified.

Reviewer: ```suggestion
The [attachments](#attachment) associated to the answer. This can be useful to allow polls where answers are represented by a media content (e.g. an image) instead of a text. Note that either one of the text or attachments field must be specified within a provided answer to make sure it's valid.
```

### End Date

> The date when the poll will close.

Reviewer: ```suggestion
The date when the poll will close. Once this date has passed, the final results of the poll will be automatically tallied on the next block after such date. Also, all new answers after the date will not be considered valid. 
```

### Allow Multiple Answers

> This field tells if the poll allows multiple answers or not.

Reviewer: ```suggestion
This field tells if the poll allows multiple answers from the same user or not.
```

### Allow Answer Edits
This field tells if the poll allows users to edit their answers or not.

### Final Tally Results 

> This fields contains the final results of the poll.

Reviewer: ```suggestion
This fields contains the final results of the poll. It's going to be populated by the chain after the poll has ended. Trying to provide a final tally results value inside a `MsgCreatePost` or `MsgAddPostAttachment` will return an error.
```

#### Results

> The answers' results represent by the `AnswerResult` structure.

Reviewer: ```suggestion
The poll results for each provided answer, represented each one by an `AnswerResult`.
```

##### Answer Result

> This field contains the result of a single poll provided answer

Reviewer: ```suggestion
This structure contains the result of a single poll provided answer
```

###### Answer Index

> The Index of the answer inside the `ProvidedAnswer`s array.

Reviewer: ```suggestion
The index of the answer within the post's `ProvidedAnswer` field associated to which these results.
```

###### Votes
The number of votes received by an answer.

## User Answer
The user answer represent an answer given by a user to a poll.

#### Subspace ID

> The [subspace] ID indicates the ID of the Dapp where the user answer is hosted and lives.

Reviewer: ```suggestion
The subspace id represents the id of the subspace inside which is placed the post that contains the poll to which this answer refers. 
```

#### Post ID

> The [post](#Post) ID to which the user answer is linked.

Reviewer: ```suggestion
Id of the post that contains the poll to which this answer refers.
```

#### Poll ID

> The ID of the Poll being answered.

Reviewer: ```suggestion
Id of the poll to which this answer refers. 
```

#### Answer Indexes

> The indexes of the answers inside the `ProvideAnswer`s array.

Reviewer: ```suggestion
The answer indexes contains a list of user answers, each one identified by the index of the chosen option within the poll's `ProvidedAnswer` array.
```
